plugin/proxy/proxy.go

```diff
@@ -97,7 +98,7 @@ func (p Proxy) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (
 				child.Finish()
 			}
 
-			taperr := toDnstap(ctx, host.Name, upstream.Exchanger(), state, reply, start)
+			taperr := dnstap.ToMessage(ctx, host.Name, upstream.Exchanger().Protocol(), state, reply, start)
```

Comment:
Isn't `upstream.Exchanger().Protocol()` returning "dns" or "https_google"?

proxy/dns.go:
~~~ Go
func (d *dnsEx) Protocol() string { return "dns" }
~~~

This is why I introduced Exchanger.Transport() returning "tcp", "udp" or "".
"" if the request used `state.Proto()`

Reply:
Fair enough, but let me rephrase: how do we want (maybe external) proxies to report dnstap? I.e. forward does not have an Exchanger?